core/src/main/java/io/undertow/server/handlers/BlockingTimeoutHandler.java

```diff
@@ -0,0 +1,367 @@
+package io.undertow.server.handlers;
+
+import io.undertow.server.ConduitWrapper;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.server.ServerConnection;
+import org.xnio.IoUtils;
+import org.xnio.XnioIoThread;
+import org.xnio.XnioWorker;
+import org.xnio.channels.ReadTimeoutException;
+import org.xnio.channels.StreamSinkChannel;
+import org.xnio.channels.StreamSourceChannel;
+import org.xnio.channels.WriteTimeoutException;
+import org.xnio.conduits.ReadReadyHandler;
+import org.xnio.conduits.StreamSinkConduit;
+import org.xnio.conduits.StreamSourceConduit;
+import org.xnio.conduits.WriteReadyHandler;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * BlockingTimeoutHandler allows configurable blocking I/O timeouts for read
+ * and write operations within an exchange.
+ * Unlike Options.READ_TIMEOUT and Options.WRITE_TIMEOUT this only applies to
+ * blocking operations which can be helpful to prevent the worker pool from
+ * becoming saturated when clients stop responding.
+ *
+ * @author Carter Kozak
+ */
+public final class BlockingTimeoutHandler implements HttpHandler {
```

Comment:
Open question: should this be split into separate BlockingReadTimeoutHandler and BlockingWriteTimeoutHandler?

Reply:
I think this makes more sense to split into separate handlers for read and write timeouts, that way each one is responsible for exactly one thing.